File: examples/scalar_law/PROGRAM1/linearad.jl
#***********************************************************************
#  Copyright 2006 John A. Trangenstein
#
#  This software is made available for research and instructional use 
#  only. 
#  You may copy and use this software without charge for these 
#  non-commercial purposes, provided that the copyright notice and 
#  associated text is reproduced on all copies.  
#  For all other uses (including distribution of modified versions), 
#  please contact the author at
#    John A. Trangenstein
#    Department of Mathematics
#    Duke University
#    Durham, NC 27708-0320
#    USA
#  or
#    johnt@math.duke.edu
#  
#  This software is made available "as is" without any assurance that it
#  is completely correct, or that it will work for your purposes.  
#  Use the software at your own risk.
#***********************************************************************

function fluxderv(fc::Int,lc::Int,ifirst::Int,ilast::Int,
                  conservd::OffsetArray{Float64,1}, dfdu::OffsetArray{Float64,1})
#   ******************************************************************
#   compute derivative of flux with respect to conserved variable
#   ******************************************************************
    @unsafe for ic=ifirst:ilast
        dfdu[ic]=velocity
    end
end

function riemann(fs::Int,ls::Int,ifirst::Int,ilast::Int,
                 left::OffsetArray{Float64,1},right::OffsetArray{Float64,1},
                 flux::OffsetArray{Float64,1})
#   ******************************************************************
#   the following riemann solver is hard-wired for linear advection
#   with velocity > 0
#   ******************************************************************
    @unsafe for ie=ifirst:ilast+1
        flux[ie]=velocity*left[ie]
    end
end 

File: examples/scalar_law/PROGRAM1/fortran/linearad.f
c***********************************************************************
c  Copyright 2006 John A. Trangenstein
c
c  This software is made available for research and instructional use 
c  only. 
c  You may copy and use this software without charge for these 
c  non-commercial purposes, provided that the copyright notice and 
c  associated text is reproduced on all copies.  
c  For all other uses (including distribution of modified versions), 
c  please contact the author at
c    John A. Trangenstein
c    Department of Mathematics
c    Duke University
c    Durham, NC 27708-0320
c    USA
c  or
c    johnt@math.duke.edu
c  
c  This software is made available "as is" without any assurance that it
c  is completely correct, or that it will work for your purposes.  
c  Use the software at your own risk.
c***********************************************************************
      subroutine fluxderv(fc,lc,ifirst,ilast,
     &  conservd,
     &  dfdu    )
c+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
      include "const.i"
      include "linearad.i"
c+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
      integer fc,lc,ifirst,ilast
      double precision conservd(fc:lc)
c
      double precision dfdu(fc:lc)
c
      integer ic
c     ******************************************************************
c     compute derivative of flux with respect to conserved variable
c     ******************************************************************
      do ic=ifirst,ilast
        dfdu(ic)=velocity
      enddo

      return
      end 

      subroutine riemann(fs,ls,ifirst,ilast,
     &  left,right,
     &  flux)
c+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
      include "linearad.i"
c+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
      integer fs,ls,ifirst,ilast
c
      double precision  
     &  left(fs:ls+1),
     &  right(fs:ls+1)
c
      double precision flux(fs:ls+1)
      integer ie
c     ******************************************************************
c     the following riemann solver is hard-wired for linear advection
c     with velocity > 0
c     ******************************************************************
      do ie=ifirst,ilast+1
        flux(ie)=velocity*left(ie)
      enddo
      return
      end 
